### PR TITLE
Document relations component refactored

### LINKED ISSUE #421: [UX] Backoffice Document details page
❓ verify first 3 with the users
- [x] define metadata to display
- [x] define all elements of the page
- [x] define types of performed actions (f.e. checking statistics, getting document data, requesting a loan, etc) to identify groups of actions 
- [x] define weight of the elements and most performed actions on this page
- [x] implement missing and rearrange elements

### PR DESCRIPTION
**Pull request by kpsherva**
(closes https://github.com/inveniosoftware/invenio-app-ils/issues/421)

Deleting:
Feb-06-2020 14-07-26

Adding:
Feb-06-2020 14-09-51

Adding with a field:
Feb-06-2020 14-12-49

* each relation has it's own component
* each component has one small task 
* additional components for titles

### CODE REVIEW

File: ui/src/components/Document/DocumentEdition.js [hunk 1]
```diff
@@ -1,5 +1,16 @@
 import React from 'react';
+import { Item } from 'semantic-ui-react';
 
-export const DocumentEdition = ({ document }) => {
-  return document.metadata.edition ? `ed.  ${document.metadata.edition}, ` : '';
+export const DocumentEdition = ({ document, explicit }) => {
+  if (explicit) {
```
Comment (ntarocco) [inline ui/src/components/Document/DocumentEdition.js]: I am not sure I understand this `explicit` param, why is that needed?

File: ui/src/components/Document/DocumentTitle.js [hunk 2]
```diff
@@ -1,11 +1,22 @@
 import { DocumentEdition } from '@components/Document';
 import React from 'react';
+import Truncate from 'react-truncate';
 
-export const DocumentTitle = ({ document }) => {
+export const DocumentTitle = ({ document, short, truncate }) => {
```
Comment (ntarocco) [inline ui/src/components/Document/DocumentTitle.js:5]: IMHO, it is tricky to play with the args like this. If I want a `short`+`truncate`? :)

I don't have great suggestions... what about something like this? (pseudo-code)
```
const DocumentTitle = ({ document, options={short: false, truncate:false } }) => {
    const title = short ? document.metadata.title : `{document.metadata.title} (<DocumentEdition document={document} /> {document.metadata.publication_year})`;
    const cmp = <div className="document-title">{title}</div>;
    return truncate ? <Truncate lines={2} ellipsis={'... '}>{cmp}</Truncate> : cmp;
```

File: ui/src/components/ESSelector/HitsSearch.js [hunk 3]
```diff
@@ -21,22 +22,27 @@ const ResultRenderer = ({ id, title, description, extra }) => (
...
 
   onSelectResult = (event, { result }) => {
     if (this.state.hasError) return;
-
     if (this.props.onSelect) {
       this.props.onSelect(result);
     }
-    this.setState(initialState);
+
+    /* controls closing of the results list in custom cases */
```
Comment (ntarocco) [inline ui/src/components/ESSelector/HitsSearch.js]: do you mind to explain a little bit better the comment? :)

File: ui/src/components/ESSelector/HitsSearch.js [hunk 4]
```diff
@@ -47,12 +53,17 @@ export class HitsSearch extends Component {
         ? searchQuery + '*'
         : searchQuery;
       const response = await this.props.query(queryString);
-      const results = [];
-      for (let hit of response.data.hits) {
-        results.push(serialize(hit));
-      }
-      if (this.props.onResults) {
-        this.props.onResults(results);
+      let results = [];
```
Comment (ntarocco) [inline ui/src/components/ESSelector/HitsSearch.js:56]: Should you call `this.props.onResults` even if `serialize` is `false` or not?

```
results = serialize ? response.data.hits.map(hit => serialize(hit)) : response.data.hits;
if (this.props.onResults) {
  this.props.onResults(results);
}
```

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/DocumentSeries.js [hunk 5]
```diff
@@ -21,33 +22,33 @@ export default class DocumentSeries extends Component {
             </>
           ),
         },
-        render: () => <RelationMultipart />,
+        render: () => (
+          <Tab.Pane className="relations-tab">
```
Comment (ntarocco) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/DocumentSeries.js]: `bo-relations-tab`? Or generic without `bo`?

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/RelationEdition/RelationEditionModal.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    }
+  };
+
+  selectResultRender = option => {
+    const existingRelations = this.props.relations;
+    const currentSelections = this.state.selections;
+
+    let disabled = false;
+    if (
+      (!isEmpty(existingRelations.edition) &&
```
Comment (ntarocco) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/RelationEdition/RelationEditionModal.js]: May I suggest to refactor it for readability? E.g.
```
const hasEdition = !isEmpty(existingRelations.edition);
const hasExistingRelation = existingRelations.edition.find(o => o.pid === option.metadata.pid));
...

const disabled = hasEdition && hasExistingRelation ...;
```

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationModal/RelationModal.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    } else {
+      selections.map(selection =>
+        newRelations.push(
+          siblingRelationPayload(relationType, extraRelationField, selection)
+        )
+      );
+    }
+
+    const pid = refererRecord.pid;
+    this.props.createRelations(pid, newRelations);
```
Comment (ntarocco) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationModal/RelationModal.js]: Why this `createRelations`? For what it is needed?

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationRemover/RelationRemover.js [hunk 8]
```diff
@@ -12,17 +13,41 @@ export default class RelationRemover extends Component {
...
 
   handleDelete = () => {
-    const { deletePayload, refererPid } = this.props;
+    const { related, referer, relationPayloadType } = this.props;
+
     this.setState({ modalOpen: false });
-    this.props.deleteRelations(refererPid, [deletePayload]);
+    let deletePayload = {};
+    if (relationPayloadType === 'siblings') {
+      deletePayload = {
```
Comment (ntarocco) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationRemover/RelationRemover.js]: I would try to avoid as much as possible to prepare a REST API `payload` in a component. I would rather pass all params to the function, and let the `api` js get the params and transform them to the needed payload.

**Review by ntarocco — COMMENTED** [review]
Nice job! Congrats! 👏 
It looks easy when reading the components implementations! :)

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationRemover/RelationRemover.js (hunk 8 above)
Comment (kpsherva) [reply ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationRemover/RelationRemover.js]: 👍 very good point

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/RelationEdition/RelationEditionModal.js (hunk 6 above)
Comment (kpsherva) [reply ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/RelationEdition/RelationEditionModal.js]: 👍

**Comment by ntarocco** [comment]
@kprzewa before we forget can you please take into account in your next PR to display, both in **details** and **search results**, if a document is restricted or not! It should be very visible. Thanks!

File: ui/src/components/ESSelector/HitsSearch.js (hunk 4 above)
Comment (kpsherva) [reply ui/src/components/ESSelector/HitsSearch.js:56]: actually I have checked and it was a leftover from something else, it is not used and I removed it.

File: ui/src/components/Document/DocumentEdition.js (hunk 1 above)
Comment (kpsherva) [reply ui/src/components/Document/DocumentEdition.js]: to display the label, I renamed it

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentEItems/DocumentEItems.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        to={BackOfficeRoutes.eitemDetailsFor(row.metadata.pid)}
+        data-test={row.metadata.pid}
+      >
+        {row.metadata.pid}
+      </Link>
+    );
+  };
+
+  filesFieldFormatter = ({ row }) => {
+    return row.metadata.files.length;
```
Comment (topless) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentEItems/DocumentEItems.js:36]: What if it doesn't have any files?

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationListEntry/RelationListEntry.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  DocumentItemCover,
+  DocumentTitle,
+} from '@components/Document';
+import { BackOfficeRoutes } from '@routes/urls';
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { Link } from 'react-router-dom';
+import { Grid, Icon, Item } from 'semantic-ui-react';
+
+export class RelationListEntry extends Component {
```
Comment (topless) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationListEntry/RelationListEntry.js:14]: Since everything is called item, like `<Item`, "item-image-wrapper" etc should we call this `RelationListItem` ? I see where this can actually become a problem when working with `items`, `eitems` though.

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationRemover/RelationRemover.js [hunk 11]
```diff
@@ -45,11 +51,12 @@ export default class RelationRemover extends Component {
 }
 
 RelationRemover.propTypes = {
-  /* Description of the relation to be deleted */
-  deletePayload: PropTypes.object.isRequired,
-  /* pid of the record using this remover */
+  /* pid of the record calling this remover */
   /* TODO PID AS OBJECT */
```
Comment (topless) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationRemover/RelationRemover.js:55]: We should add an issue for that

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationSelector/state/reducer.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  switch (action.type) {
+    case SELECT_OPTION:
+      let newSelections = [...state.selections];
+      newSelections.push(action.option);
+      return {
+        selections: newSelections,
+      };
+    case REMOVE_SELECTION:
+      let removeSelections = [...state.selections];
+      remove(removeSelections, o => o.metadata.pid === action.removePid);
```
Comment (topless) [inline ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationSelector/state/reducer.js:18]: I am not sure this is the place to manipulate data.

File: ui/src/pages/backoffice/DocumentRequest/DocumentRequestSearch/DocumentRequestListEntry.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            as={Link}
+            to={BackOfficeRoutes.documentRequestDetailsFor(
+              documentRequest.metadata.pid
+            )}
+            data-test={`navigate-${documentRequest.metadata.pid}`}
+          >
+            <DocumentRequestIcon /> Literature request #
+            {documentRequest.metadata.pid}
+          </Item.Header>
+          <Grid columns={5}>
```
Comment (topless) [inline ui/src/pages/backoffice/DocumentRequest/DocumentRequestSearch/DocumentRequestListEntry.js:24]: It says `5` columns but I count only `3`

**Review by topless — APPROVED** [review]
Massive, as always 🎉 !! Looks great, and looks also ready to accommodate the new `sequence` relation!

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentEItems/DocumentEItems.js (hunk 9 above)
Comment (kpsherva) [reply ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentEItems/DocumentEItems.js:36]: the array is always there, it will just show 0

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationListEntry/RelationListEntry.js (hunk 10 above)
Comment (kpsherva) [reply ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationListEntry/RelationListEntry.js:14]: there is also `<Item>` as semantic-ui component, I didn't want to mix the names

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationSelector/state/reducer.js (hunk 12 above)
Comment (kpsherva) [reply ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationSelector/state/reducer.js:18]: if you know where I can manipulate the state like this I could change it, but I didn't have a better idea
@ntarocco WDYT?

File: ui/src/pages/backoffice/DocumentRequest/DocumentRequestSearch/DocumentRequestListEntry.js (hunk 13 above)
Comment (kpsherva) [reply ui/src/pages/backoffice/DocumentRequest/DocumentRequestSearch/DocumentRequestListEntry.js:24]: I will fix it in the next PR

File: ui/src/api/documents/document.js [hunk 14]
```diff
@@ -33,16 +38,75 @@ const viewEvent = async docPid => {
...
+      parent_pid_type: related.pid_type,
+      child_pid: referer.metadata.pid,
+      child_pid_type: recordToPidType(referer),
+      relation_type: related.relation_type,
+    };
+  }
+  const resp = await http.delete(
+    `${documentURL}${referer.metadata.pid}/relations`,
+    {
+      data: [deleteRequestPayload],
```
Comment (ntarocco) [inline ui/src/api/documents/document.js:107]: do we need to pass a list here?

File: ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationSelector/state/reducer.js (hunk 12 above)
Comment (ntarocco) [reply ui/src/pages/backoffice/Document/DocumentDetails/components/DocumentRelations/components/RelationSelector/state/reducer.js:18]: @topless what do you mean exactly? The reducer should be the place where to change the state, no?
See [here](https://redux.js.org/basics/reducers/#handling-more-actions).
Do you maybe have something else in mind?

File: invenio_app_ils/records_relations/api.py [hunk 15]
```diff
@@ -180,6 +180,8 @@ def _build_sibling(self, related_pid, relation_type):
 
         # add also title, language and edition of the sibling
         r["title"] = sibling.get("title", "")
+        r["document_type"] = sibling.get("document_type", "")
```
Comment (ntarocco) [inline invenio_app_ils/records_relations/api.py:183]: I think I did something a bit silly previously here, I was adding `title` in any case, even if empty, but not language or edition if empty.
We should probably decide one way :)

**Review by ntarocco — APPROVED** [review]
I went quite fast on this, we can eventually iterate more later on.

File: ui/src/api/documents/document.js [hunk 16]
```diff
@@ -33,16 +38,75 @@ const viewEvent = async docPid => {
...
+  if (relationType === 'serial' || relationType === 'multipart_monograph') {
+    const payload = parentChildRelationPayload(
+      relationType,
+      extraRelationField,
+      selectedRelatedList[0],
+      referer
+    );
+    newRelationsPayload.push(payload);
+  } else {
+    if (relationType === 'other') {
```
[the remark this answers is not in the document]
Comment (kpsherva) [reply ui/src/api/documents/document.js:57]: I will fix in the next PR

File: ui/src/api/documents/document.js (hunk 14 above)
Comment (kpsherva) [reply ui/src/api/documents/document.js:107]: I didn't want to change the api on the backend until I will also redo the series relations
